common/cache/simple.go

```diff
@@ -0,0 +1,212 @@
+	// make a copy of the entry so there will be no concurrent access to this entry
+	entry = &simpleEntry{
+		key:   entry.key,
+		value: entry.value,
+	}
+	return entry
+}
+
+func (e *simpleEntry) Key() interface{} {
+	return e.key
+}
+
+func (e *simpleEntry) Value() interface{} {
+	return e.value
+}
+
+// CreateTime is not implemented for simple cache entries
+func (e *simpleEntry) CreateTime() time.Time {
+	return DummyCreateTime
+}
+
+// NewSimple creates a new simple cache with given options.
+// Simple cache will never evict entries and it will never reorder the elements.
+// Simple cache also does not have the concept of pinning that LRU cache has.
+// Internally simple cache uses a RWMutex instead of the exclusive Mutex that LRU cache uses.
+// The RWMutex makes simple cache readable by many threads without introducing lock contention.
+// The maxSize provided to simple cache should be larger than the possible largest size because
+// simple cache will simply panic if max size is exceeded.
+func NewSimple(maxSize int, opts *SimpleOptions) Cache {
+	if opts == nil {
+		opts = &SimpleOptions{}
+	}
+	return &simple{
+		iterateList: list.New(),
+		accessMap:   make(map[interface{}]*list.Element, opts.InitialCapacity),
+		maxSize:     maxSize,
+		rmFunc:      opts.RemovedFunc,
+	}
+}
+
+// Get retrieves the value stored under the given key
+func (c *simple) Get(key interface{}) interface{} {
+	c.RLock()
+	defer c.RUnlock()
+
+	element := c.accessMap[key]
+	if element == nil {
+		return nil
+	}
+	return element.Value.(*simpleEntry).Value()
+}
+
+// Put puts a new value associated with a given key, returning the existing value (if present).
+// If Put is called on a cache which is already full it will panic.
+func (c *simple) Put(key interface{}, value interface{}) interface{} {
+	c.Lock()
+	defer c.Unlock()
+	existing, err := c.putInternal(key, value, true)
+	if err != nil {
+		panic(err)
+	}
+	return existing
+}
+
+// PutIfNotExist puts a value associated with a given key if it does not exist
+func (c *simple) PutIfNotExist(key interface{}, value interface{}) (interface{}, error) {
+	c.Lock()
+	defer c.Unlock()
+	existing, err := c.putInternal(key, value, false)
+	if err != nil {
+		return nil, err
+	}
+	if existing == nil {
+		// This is a new value
+		return value, nil
+	}
+	return existing, nil
+}
+
+// Delete deletes a key, value pair associated with a key
+func (c *simple) Delete(key interface{}) {
+	c.Lock()
+	defer c.Unlock()
+
+	element := c.accessMap[key]
+	if element == nil {
+		return
+	}
+	entry := c.iterateList.Remove(element).(*simpleEntry)
+	if c.rmFunc != nil {
+		go c.rmFunc(entry.value)
+	}
+	delete(c.accessMap, entry.key)
+}
+
+// Release does nothing for simple cache
+func (c *simple) Release(_ interface{}) {}
+
+// Size returns the number of entries currently in the cache
+func (c *simple) Size() int {
+	c.RLock()
+	defer c.RUnlock()
+
+	return len(c.accessMap)
+}
+
+func (c *simple) Iterator() Iterator {
```

Comment:
The current way could work. Could also just return a channel. So we are not leaking the lock to the caller. Also can get rid of the list.

```
func (c *simple) Iterator() chan<- Entry {
	iterChan := make(chan Entry, 1000)

	go func() {
		c.RLock()
		defer c.RUnlock()
		for key, value := range c.accessMap {
			iterChan <- &simpleEntry{
				key:   key,
				value: value,
			}
		}
		close(iterChan)
	}()

	return iterChan
}
```

Reply:
We cannot unless we change the interface for Cache which will impact LRU cache as well - pretty big change.

Reply:
I know. Not suggesting you to change it in this PR.